### PR TITLE
fix(@angular/ssr): improve handling of route mismatches between Angular server routes and Angular router

### PR DESCRIPTION
**Pull request by alan-agius4**
fix(@angular/ssr): show error when multiple routes are set with `RenderMode.AppShell`**

This change introduces error handling to ensure that when multiple routes are configured with `RenderMode.AppShell`, an error message is displayed. This prevents misconfiguration and enhances clarity in route management.

---

**fix(@angular/ssr): improve handling of route mismatches between Angular server routes and Angular router**

This commit resolves an issue where routes defined in the Angular server routing configuration did not match those in the Angular router. Previously, discrepancies between these routes went unnoticed by users. With this update, appropriate error messages are now displayed when mismatches occur, enhancing the developer experience and facilitating easier troubleshooting.

### CODE REVIEW

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 1]
```diff
@@ -43,7 +43,10 @@ const VALID_REDIRECT_RESPONSE_CODES = new Set([301, 302, 303, 307, 308]);
 /**
  * Additional metadata for a server configuration route tree.
  */
-type ServerConfigRouteTreeAdditionalMetadata = Partial<ServerRoute>;
+type ServerConfigRouteTreeAdditionalMetadata = Partial<ServerRoute> & {
+  /** Indicates if the route has been matched with the Angular router routes. */
+  matched?: boolean;
```
Comment (AndrewKushnir) [inline packages/angular/ssr/src/routes/ng-routes.ts:48]: nit: I'd propose a more descriptive name for readability and easier maintenance:
```suggestion
  presentInClientRouter?: boolean;
```

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 2]
```diff
@@ -124,19 +127,23 @@ async function* traverseRoutesConfig(options: {
...
+
+        matchedMetaData.matched = true;
       }
 
       const metadata: ServerConfigRouteTreeNodeMetadata = {
         ...matchedMetaData,
         route: currentRoutePath,
       };
 
+      delete metadata.matched;
```
Comment (AndrewKushnir) [inline packages/angular/ssr/src/routes/ng-routes.ts:145]: The `delete` operation used to cause deopt in VM, may be use `null` or `undefined` instead?
```suggestion
      metadata.matched = null;
```

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 3]
```diff
@@ -433,13 +442,40 @@ export async function getRoutesFromAngularRouterConfig(
...
 
+      let seenAppShellRoute: string | undefined;
       for await (const result of traverseRoutes) {
         if ('error' in result) {
           errors.push(result.error);
         } else {
+          if (result.renderMode === RenderMode.AppShell) {
+            if (seenAppShellRoute !== undefined) {
+              errors.push(
+                `Error: Both the route '${seenAppShellRoute}' and '${stripLeadingSlash(result.route)}' have ` +
```
Comment (AndrewKushnir) [inline packages/angular/ssr/src/routes/ng-routes.ts:453]:
```suggestion
                `Error: Both the '${seenAppShellRoute}' and '${stripLeadingSlash(result.route)}' routes have ` +
```

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 4]
```diff
@@ -433,13 +442,40 @@ export async function getRoutesFromAngularRouterConfig(
...
+
+      if (serverConfigRouteTree) {
+        for (const { route, matched, renderMode } of serverConfigRouteTree.traverse()) {
+          if (matched || route === '**') {
+            // Skip if matched or it's the catch-all route.
+            continue;
+          }
+
+          errors.push(
+            `The server route '${route}' does not match any routes defined in the Angular routing configuration. ` +
```
Comment (AndrewKushnir) [inline packages/angular/ssr/src/routes/ng-routes.ts:474]:
```suggestion
            `The '${route}' server route does not match any routes defined in the Angular routing configuration (typically provided as a part of the \`provideRouter\` call). ` +
```

File: packages/angular/ssr/src/routes/ng-routes.ts [hunk 5]
```diff
@@ -433,13 +442,40 @@ export async function getRoutesFromAngularRouterConfig(
...
+              'Please verify and if unneeded remove this route from the server configuration.',
```
Comment (AndrewKushnir) [inline packages/angular/ssr/src/routes/ng-routes.ts:475]:
```suggestion
              'Please make sure that the mentioned server route is present in the main client routing configuration.',
```

**Review by AndrewKushnir — APPROVED** [review]
LGTM (just left a few minor comments) 👍

File: packages/angular/ssr/src/routes/ng-routes.ts (hunk 1 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/ng-routes.ts:48]: This is done during build time so I am less worried about that, if we set undefined or null, we’d need to filter it out so not to include it in the manifest

File: packages/angular/ssr/src/routes/ng-routes.ts (hunk 2 above)
Comment (alan-agius4) [reply packages/angular/ssr/src/routes/ng-routes.ts:145]: This is done during build time so I am less worried about that, if we set undefined or null, we’d need to filter it out so not to include it in the manifest

**Pull request merged**